File: api/models/template.py
from pydantic import BaseModel, Field
from typing import List, Dict, Optional, Any, Union
from datetime import datetime
from enum import Enum

class TemplateStatus(str, Enum):
    """Status of a template"""
    DRAFT = "Draft"
    PUBLISHED = "Published"
    ARCHIVED = "Archived"

class TemplateType(str, Enum):
    """Type of template"""
    PAGE = "Page"
    SECTION = "Section"
    COMPONENT = "Component"
    EMAIL = "Email"
    SOCIAL = "Social"

class FieldType(str, Enum):
    """Type of field in a template"""
    TEXT = "Text"
    RICH_TEXT = "RichText"
    NUMBER = "Number"
    BOOLEAN = "Boolean"
    DATE = "Date"
    IMAGE = "Image"
    FILE = "File"
    REFERENCE = "Reference"
    SELECT = "Select"
    MULTI_SELECT = "MultiSelect"
    COLOR = "Color"
    URL = "URL"
    EMAIL = "Email"
    PHONE = "Phone"
    LOCATION = "Location"
    JSON = "JSON"
    MARKDOWN = "Markdown"
    CODE = "Code"

class FieldValidation(BaseModel):
    """Validation rules for a field"""
    required: bool = False
    min_length: Optional[int] = None
    max_length: Optional[int] = None
    min_value: Optional[float] = None
    max_value: Optional[float] = None
    pattern: Optional[str] = None
    options: Optional[List[str]] = None
    default_value: Optional[Any] = None

class TemplateField(BaseModel):
    """Field definition for a template"""
    id: str
    name: str
    description: Optional[str] = None
    type: FieldType
    validation: Optional[FieldValidation] = None
    is_localized: bool = False
    is_searchable: bool = False
    is_hidden: bool = False
    group: Optional[str] = None
    order: int = 0

class TemplateSection(BaseModel):
    """Section in a template"""
    id: str
    name: str
    description: Optional[str] = None
    fields: List[TemplateField]
    is_repeatable: bool = False
    min_items: Optional[int] = None
    max_items: Optional[int] = None
    order: int = 0

class TemplateCreate(BaseModel):
    """Model for creating a new template"""
    name: str
    description: Optional[str] = None
    type: TemplateType
    sections: List[TemplateSection]
    status: TemplateStatus = TemplateStatus.DRAFT
    is_default: bool = False
    thumbnail_url: Optional[str] = None
    tags: Optional[List[str]] = None
    metadata: Optional[Dict[str, Any]] = None

class TemplateUpdate(BaseModel):
    """Model for updating a template"""
    name: Optional[str] = None
    description: Optional[str] = None
    type: Optional[TemplateType] = None
    sections: Optional[List[TemplateSection]] = None
    status: Optional[TemplateStatus] = None
    is_default: Optional[bool] = None
    thumbnail_url: Optional[str] = None
    tags: Optional[List[str]] = None
    metadata: Optional[Dict[str, Any]] = None

class TemplateVersion(BaseModel):
    """Version of a template"""
    version: int
    created_at: datetime
    created_by: str
    changes: Optional[str] = None
    template_data: Dict[str, Any]

class TemplateResponse(BaseModel):
    """Response model for template information"""
    id: str
    name: str
    description: Optional[str] = None
    type: TemplateType
    sections: List[TemplateSection]
    status: TemplateStatus
    is_default: bool = False
    thumbnail_url: Optional[str] = None
    tags: Optional[List[str]] = None
    metadata: Optional[Dict[str, Any]] = None
    version: int = 1
    created_at: datetime
    updated_at: datetime
    created_by: str
    updated_by: str
    
    class Config:
        from_attributes = True

File: src/types/template.ts
/**
 * Template status enum
 */
export enum TemplateStatus {
  DRAFT = 'Draft',
  PUBLISHED = 'Published',
  ARCHIVED = 'Archived'
}

/**
 * Template type enum
 */
export enum TemplateType {
  PAGE = 'Page',
  SECTION = 'Section',
  COMPONENT = 'Component',
  EMAIL = 'Email',
  SOCIAL = 'Social'
}

/**
 * Field type enum
 */
export enum FieldType {
  TEXT = 'Text',
  RICH_TEXT = 'RichText',
  NUMBER = 'Number',
  BOOLEAN = 'Boolean',
  DATE = 'Date',
  IMAGE = 'Image',
  FILE = 'File',
  REFERENCE = 'Reference',
  SELECT = 'Select',
  MULTI_SELECT = 'MultiSelect',
  COLOR = 'Color',
  URL = 'URL',
  EMAIL = 'Email',
  PHONE = 'Phone',
  LOCATION = 'Location',
  JSON = 'JSON',
  MARKDOWN = 'Markdown',
  CODE = 'Code'
}

/**
 * Field validation interface
 */
export interface FieldValidation {
  required?: boolean;
  min_length?: number;
  max_length?: number;
  min_value?: number;
  max_value?: number;
  pattern?: string;
  options?: string[];
  default_value?: any;
}

/**
 * Template field interface
 */
export interface TemplateField {
  id: string;
  name: string;
  description?: string;
  type: string;
  validation?: FieldValidation;
  is_searchable?: boolean;
  is_localized?: boolean;
  is_hidden?: boolean;
  group?: string;
  order: number;
}

/**
 * Template section interface
 */
export interface TemplateSection {
  id: string;
  name: string;
  description?: string;
  fields: TemplateField[];
  is_repeatable: boolean;
  min_items?: number;
  max_items?: number;
  order: number;
}

/**
 * Template create interface
 */
export interface TemplateCreate {
  name: string;
  description?: string;
  type: string;
  sections: TemplateSection[];
  status: string;
  is_default: boolean;
  thumbnail_url?: string;
  tags?: string[];
  metadata?: Record<string, any>;
}

/**
 * Template update interface
 */
export interface TemplateUpdate {
  name?: string;
  description?: string;
  type?: string;
  sections?: TemplateSection[];
  status?: string;
  is_default?: boolean;
  thumbnail_url?: string;
  tags?: string[];
  metadata?: Record<string, any>;
}

/**
 * Template version interface
 */
export interface TemplateVersion {
  version: number;
  created_at: string;
  created_by: string;
  changes?: string;
  template_data: Record<string, any>;
}

/**
 * Template response interface
 */
export interface TemplateResponse {
  id: string;
  name: string;
  description?: string;
  type: string;
  sections: TemplateSection[];
  status: string;
  is_default: boolean;
  thumbnail_url?: string;
  tags?: string[];
  metadata?: Record<string, any>;
  version: number;
  created_at: string;
  updated_at: string;
  created_by: string;
  updated_by: string;
}
